lib/runtime/wasmer/imports.go

```diff
@@ -185,45 +188,191 @@ func ext_sandbox_memory_teardown_version_1(context unsafe.Pointer, a C.int32_t)
+		return 0
+	}
+
+	ret, err := toWasmMemoryOptional(instanceContext, sig)
+	if err != nil {
+		logger.Error("[ext_crypto_ed25519_sign_version_1] failed to allocate memory", err)
+		return 0
+	}
+
+	return C.int64_t(ret)
 }
 
 //export ext_crypto_ed25519_verify_version_1
-func ext_crypto_ed25519_verify_version_1(context unsafe.Pointer, a C.int32_t, z C.int64_t, y C.int32_t) C.int32_t {
+func ext_crypto_ed25519_verify_version_1(context unsafe.Pointer, sig C.int32_t, msg C.int64_t, key C.int32_t) C.int32_t {
 	logger.Trace("[ext_crypto_ed25519_verify_version_1] executing...")
-	logger.Warn("[ext_crypto_ed25519_verify_version_1] unimplemented")
-	return 0
+
+	instanceContext := wasm.IntoInstanceContext(context)
+	memory := instanceContext.Memory().Data()
+	signVerify := instanceContext.Data().(*runtime.Context).SigVerifier
+
+	newSign := memory[sig : sig+64]
+	newMsg := asMemorySlice(instanceContext, msg)
+	pubKeyData := memory[key : key+32]
+
+	pubKey, err := ed25519.NewPublicKey(pubKeyData)
+	if err != nil {
+		logger.Error("[ext_crypto_ed25519_verify_version_1] failed to access public Key")
+		return 0
+	}
+
+	if signVerify.IsStarted() {
+		signature := runtime.Signature{
+			PubKey:    pubKey.Encode(),
+			Sign:      newSign,
+			Msg:       newMsg,
+			KyeTypeID: crypto.Ed25519Type,
+		}
+		signVerify.Add(&signature)
+		return 0
+	}
+
+	if ok, err := pubKey.Verify(newMsg, newSign); err != nil || !ok {
+		logger.Error("[ext_crypto_ed25519_verify_version_1] failed to verify")
+		return 0
+	}
+
+	return 1
 }
 
 //export ext_crypto_finish_batch_verify_version_1
 func ext_crypto_finish_batch_verify_version_1(context unsafe.Pointer) C.int32_t {
 	logger.Trace("[ext_crypto_finish_batch_verify_version_1] executing...")
-	logger.Warn("[ext_crypto_finish_batch_verify_version_1] unimplemented")
-	return 1
+
+	instanceContext := wasm.IntoInstanceContext(context)
+	signVerify := instanceContext.Data().(*runtime.Context).SigVerifier
+
+	if !signVerify.IsStarted() {
+		logger.Error("[ext_crypto_finish_batch_verify_version_1] batch verification is not started")
+		panic("batch verification is not started")
+	}
+
+	if signVerify.Finish() {
+		return 1
+	}
+	return 0
 }
 
 //export ext_crypto_secp256k1_ecdsa_recover_version_1
-func ext_crypto_secp256k1_ecdsa_recover_version_1(context unsafe.Pointer, a, z C.int32_t) C.int64_t {
+func ext_crypto_secp256k1_ecdsa_recover_version_1(context unsafe.Pointer, sig, msg C.int32_t) C.int64_t {
 	logger.Trace("[ext_crypto_secp256k1_ecdsa_recover_version_1] executing...")
-	logger.Warn("[ext_crypto_secp256k1_ecdsa_recover_version_1] unimplemented")
-	return 0
+	instanceContext := wasm.IntoInstanceContext(context)
+	memory := instanceContext.Memory().Data()
+
+	// msg must be the 32-byte hash of the message to be signed.
+	// sig must be a 65-byte compact ECDSA signature containing the
+	// recovery id as the last element
+	newMsg := memory[msg : msg+32]
+	newSig := memory[sig : sig+65]
+
+	// TODO: Verify the return in spec.
```

Comment:
I believe the `Result` type returns `byte(0) + key` if success, `byte(1)` otherwise, but I will confirm with the spec team

Reply:
can you update the return type for this func to be `Result`? you might need to make a `toWasmMemoryResult` func or something like that